### PR TITLE
feat(logging): add logger minimal severity

### LINKED ISSUE #6345: logging: add logger minimal severity
**Problem:**
All logs are sent to the Stackdriver, including debug logs. This is a problem for the security and privacy of logged data. Debug logs may contain sensitive data (for example: credit cards numbers or incoming request auth token).

**Solution:**
I suggest to add logger option `MinimalSeverity` (default value `Default`). `Logger.Log` and `Logger.LogSync` methods compare log entry severity with minimal severity. If entry severity lesser than minimal severity then log entry is ignored.

**Considered alternatives:**
I considered the option of creating my own wrapper over the logger, which would implement the above logic. But this logic is implemented in many well-known logging libraries (for example: zap or logrus), I decided to try to create this issue.

### PR DESCRIPTION
**Pull request by kimmelserj**
Fixes #6345

### CODE REVIEW

File: logging/loggeroption.go
Comment (minherz) [inline logging/loggeroption.go:193]: nit: I'd suggest to make the name more verbose. "MinimalSeverity" requires additional clarification. Consider "IngestedSeverityThreshold" or "MinimalLoggedSeverity" or something similar.

File: logging/logging.go
Comment (minherz) [inline logging/logging.go:633]: nit: why not `<=`? Inclusive threshold is implied by the name: "minimal severity" -> "severity below which nothing is ingested"
nit: should it return a specialized error indicating that the ingestion did not take place? returning `nil` means everything is OK but the job still was not done.

File: logging/logging_test.go
Comment (minherz) [inline logging/logging_test.go:328]: nit: it looks like a great opportunity for refactoring since both `TestLogSyncWithMinimalSeverity()` and `TestLogWithMinimalSeverity()` have very similar code lines.

**Review by minherz — COMMENTED** [review]
Please, have a look at provided comments.

**Comment by minherz** [comment]
@daniel-sanche please have a look into this new contribution.

File: logging/loggeroption.go
Comment (kimmelserj) [reply logging/loggeroption.go:193]: Ok. I will rename `MinimalSeverity` to `MinimalLoggedSeverity`.

File: logging/logging.go
Comment (kimmelserj) [reply logging/logging.go:633]: > nit: why not <=? Inclusive threshold is implied by the name: "minimal severity" -> "severity below which nothing is ingested"

I came from the definition of a minimum: https://en.wikipedia.org/wiki/Maxima_and_minima. The minimal severity is the minimum allowed severity value. That's why I used `<` in the expression.

> nit: should it return a specialized error indicating that the ingestion did not take place? returning nil means everything is OK but the job still was not done.

An error is defined as a difference between the desired and actual behavior. The minimal severity has desired behavior, that's why nil is returned.

File: logging/logging_test.go
Comment (kimmelserj) [reply logging/logging_test.go:252]: Got it. I will rewrite the tests to check edge cases.

File: logging/logging_test.go
Comment (kimmelserj) [reply logging/logging_test.go:328]: Ok. I will try to do something with it.

File: logging/loggeroption.go
Comment (kimmelserj) [reply logging/loggeroption.go]: Done

File: logging/logging_test.go
Comment (kimmelserj) [reply logging/logging_test.go]: Done

File: logging/logging_test.go
Comment (kimmelserj) [reply logging/logging_test.go]: Done

**Comment by kimmelserj** [comment]
@daniel-sanche could you take a look on the PR? I have already done all the suggested changes.

**Comment by kimmelserj** [comment]
@daniel-sanche 🙏

**Comment by daniel-sanche** [comment]
I definitely agree that this is an issue that should be addressed, thanks for putting this solution together!

That said, I want to be careful to make sure that this is the right solution before merging it. I'd like to discuss this with some of the go experts within GCP libraries to make sure I'm not missing any other solutions that should be considered. Since we're getting near the release freeze for the holidays, these discussions will likely have to wait for next quarter.

Sorry for the late response. Just wanted to give a quick update that this is on our radar

File: logging/logging.go
Comment (daniel-sanche) [inline logging/logging.go:257]: Is this a breaking change? Could this have any affects on existing users?

File: logging/logging.go
Comment (daniel-sanche) [inline logging/logging.go:278]: This should be mentioned with the [other LoggerOptions in the README](https://github.com/googleapis/google-cloud-go/blob/56f4cdd066cc9eaeece2c6fb466d58c3e7c41563/logging/README.md#logger-configuration-options)

**Review by daniel-sanche — CHANGES_REQUESTED** [review]
Sorry about the wait. I left a couple comments, but overall this looks like a great addition.

If you don't have the bandwidth to finish this up anymore, I can take it on too

Thanks!

File: logging/logging.go
Comment (losalex) [inline logging/logging.go:635]: Not sure if this is valid concern, but since `Default` the lowest severity, we should also introduce `None` to be the lowest severity, thus anyone could also log with `Default` severity which correct me if I wrong a valid cloud Logging severity.

File: logging/logging.go
Comment (kimmelserj) [reply logging/logging.go:257]: No. Severity zero value is zero = LogSeverity_DEFAULT. That's why [this if](https://github.com/googleapis/google-cloud-go/pull/6346/files#diff-6bb8ce03ad8192446d441a3d85bf0fb8307e3d7829bc027785b35ce95868da70R664) will be skipped if minimalLoggedSeverity is not set. It's backward compatible change.

File: logging/logging.go
Comment (kimmelserj) [reply logging/logging.go:278]: Sure. I will add this to README.

File: logging/logging.go
Comment (kimmelserj) [reply logging/logging.go:635]: I got you. The `Default` severity level is a very strange thing. [Here](https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#logseverity), the `Default` severity is described as "The log entry has no assigned severity level." We cannot recognize the severity level for "default" logs. Apparently, "default" logs will always have to be logged, regardless of the minimum severity level. I will prepare the changes in the code. Thank you for insight.)

File: logging/logging.go
Comment (kimmelserj) [reply logging/logging.go:278]: Done.

File: logging/logging.go
Comment (kimmelserj) [reply logging/logging.go]: Done.

Comment (losalex) [reply logging/logging.go]: Thanks @kimmelserj! 
@daniel-sanche, thinking about it twice I wonder if we could add Severity.None (-1) same way as we did in Java. This way we could make a logic to be straightforward - the only addition would be to make sure that this severity cannot be used in log entries

File: logging/logging.go
Comment (daniel-sanche) [inline logging/logging.go:278]: nit: A comment explaining that default means no filtering would be helpful

File: logging/logging.go
Comment (daniel-sanche) [inline logging/logging.go:635]: nit: a comment here would also be useful. Something like `Ignore logs with severity less than minimalLoggedSeverity. Default severity logs are always written`

File: logging/logging.go
Comment (daniel-sanche) [reply logging/logging.go]: > Not sure if this is valid concern, but since Default the lowest severity, we should also introduce All to be the "include-all" severity, thus anyone could also log with Default severity which correct me if I wrong a valid cloud Logging severity.

My comment is out of date now since I'm late, but I believe the original code with `minimalLoggedSeverity` was already the "include-all" option, since it was filtering logs out using `<` and not `<=`

> Apparently, "default" logs will always have to be logged, regardless of the minimum severity level. I will prepare the changes in the code. Thank you for insight.)

Hmm I'm not sure if I agree with this behaviour. I'd probably say we should treat Default as as a filterable level below debugging with severity 0 like [in the spec](https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#logseverity). But there's definitely an argument to this change too, so I'm fine either way

> @daniel-sanche, thinking about it twice I wonder if we could add Severity.None (-1) same way as we did in Java. This way we could make a logic to be straightforward - the only addition would be to make sure that this severity cannot be used in log entries

Yeah, it may be something worth considering following Java for the next step if people ask for it, but I'd vote we should keep this PR simple for now and merge this without it.

File: logging/logging.go
Comment (daniel-sanche) [inline logging/logging.go:664]: nit: same comment here too

**Review by daniel-sanche — APPROVED** [review]
LGTM with small comments, but we should wait for Alex's feedback before merging

File: logging/logging.go
Comment (losalex) [reply logging/logging.go]: > > Not sure if this is valid concern, but since Default the lowest severity, we should also introduce All to be the "include-all" severity, thus anyone could also log with Default severity which correct me if I wrong a valid cloud Logging severity.
> 
> My comment is out of date now since I'm late, but I believe the original code with `minimalLoggedSeverity` was already the "include-all" option, since it was filtering logs out using `<` and not `<=`
> 
> > Apparently, "default" logs will always have to be logged, regardless of the minimum severity level. I will prepare the changes in the code. Thank you for insight.)
> 
> Hmm I'm not sure if I agree with this behaviour. I'd probably say we should treat Default as as a filterable level below debugging with severity 0 like [in the spec](https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#logseverity). But there's definitely an argument to this change too, so I'm fine either way
> 
> > @daniel-sanche, thinking about it twice I wonder if we could add Severity.None (-1) same way as we did in Java. This way we could make a logic to be straightforward - the only addition would be to make sure that this severity cannot be used in log entries
> 
> Yeah, it may be something worth considering following Java for the next step if people ask for it, but I'd vote we should keep this PR simple for now and merge this without it.

@kimmelserj , seems adding Severity.None is not a big of a fix - can you please let us know if you can make it?

File: logging/logging_test.go
Comment (minherz) [inline logging/logging_test.go:212]: nit: consider refactoring instead of creating just another copy of the same code.

File: logging/logging_test.go
Comment (minherz) [inline logging/logging_test.go:295]: nit: same comment. the current implementation increases technical debt for tests instead of decreasing it.

**Comment by codyoss** [comment]
Closing this PR as it is outdated. Feel free to reopen if you still want to address the open feedback.

**Pull request closed without merging**
